kernel/memory_structs/src/lib.rs

```diff
@@ -294,59 +331,134 @@ macro_rules! implement_page_frame {
+                pub const fn containing_address_2mb(addr: $address) -> $TypeName<Page2MiB> {
+                    $TypeName {
+                        number: addr.value() / (PAGE_SIZE * Page2MiB::NUM_4K_PAGES) * Page2MiB::NUM_4K_PAGES,
+                        size: PhantomData,
+                    }
+                }
+            }
+            impl $TypeName<Page1GiB> {
+                #[doc = "Returns a 1GiB huge `" $TypeName "` containing the given `" $address "`."]
+                pub const fn containing_address_1gb(addr: $address) -> $TypeName<Page1GiB> {
+                    $TypeName {
+                        number: addr.value() / (PAGE_SIZE * Page1GiB::NUM_4K_PAGES) * Page1GiB::NUM_4K_PAGES,
+                        size: PhantomData,
+                    }
+                }
+            }
+            impl<P: PageSize + 'static> $TypeName<P> {
+                #[doc = "Returns the number of this `" $TypeName "`."]
+                #[inline(always)]
+                pub const fn number(&self) -> usize {
+                    self.number
+                }
+
+                #[doc = "Returns the `" $address "` at the start of this `" $TypeName "`."]
+                pub const fn start_address(&self) -> $address {
+                    $address::new_canonical(self.number * PAGE_SIZE)
+                }
+
+                #[doc = "Returns the size of this `" $TypeName "`."]
+                pub const fn page_size(&self) -> MemChunkSize {
+                    P::SIZE
+                }
+            }
+            impl<P: PageSize + 'static> fmt::Debug for $TypeName<P> {
                 fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
                     write!(f, concat!(stringify!($TypeName), "(", $prefix, "{:#X})"), self.start_address())
                 }
             }
-            impl Add<usize> for $TypeName {
-                type Output = $TypeName;
-                fn add(self, rhs: usize) -> $TypeName {
+            impl<P: PageSize + 'static> Add<usize> for $TypeName<P> {
+                type Output = $TypeName<P>;
+                fn add(self, rhs: usize) -> $TypeName<P> {
                     // cannot exceed max page number (which is also max frame number)
                     $TypeName {
-                        number: core::cmp::min(MAX_PAGE_NUMBER, self.number.saturating_add(rhs)),
+                        number: core::cmp::min(MAX_PAGE_NUMBER, self.number.saturating_add(rhs * P::NUM_4K_PAGES)),
+                        size: self.size,
                     }
                 }
             }
-            impl AddAssign<usize> for $TypeName {
+            impl<P: PageSize + Copy + 'static> AddAssign<usize> for $TypeName<P> {
                 fn add_assign(&mut self, rhs: usize) {
-                    *self = $TypeName {
-                        number: core::cmp::min(MAX_PAGE_NUMBER, self.number.saturating_add(rhs)),
-                    };
+                    *self = (*self).add(rhs * P::NUM_4K_PAGES)
```

Comment:
My bad. If we want to reduce the trait bounds for impl blocks to a minimum, I can adopt the strategy that doesn't require `Copy` for both.

Reply:
no worries, it's up to you. The `Copy` bound isn't really a problem here since we only implement the `PageSize` trait for our own 3 marker types, which are all `Copy` anyway.